proposals/0698-qualified-literals.rst

```diff
@@ -0,0 +1,565 @@
+
+::
+
+  -- only called on literals, can be used with any numeric literal: naturals, integers, rationals
+  fromNumeric :: Real a => a -> BigDecimal
+  fromNumeric = BigDecimal . realToFrac
+
+Multiple numeric types
+~~~~~~~~~~~~~~~~~~~~~~
+
+Here's an example of an implementation that's generic to all three numeric types:
+
+::
+
+  module StringInc (fromNumeric) where
+
+  fromNumeric :: (Num a, Show a) => a -> String
+  fromNumeric x = "StringInc (" ++ show (x + 1) ++ ")"
+
+* ``StringInc.10`` => ``"StringInc (11)"``
+* ``StringInc.(-10)`` => ``"StringInc (-9)"``
+* ``StringInc.(10.5)`` => ``"StringInc (23 % 2)"``
+
+Here's an example of an implementation that's different for each numeric type:
+
+::
+
+  module MyMod (fromNumeric) where
+
+  class Impl a where
+    fromNumeric :: a -> String
+
+  instance Impl Natural where
+    fromNumeric n = "NAT:" ++ show n
+
+  instance Impl Integer where
+    fromNumeric n = "NEGATIVE:" ++ show n
+
+  instance Impl Rational where
+    fromNumeric n = "FLOAT:" ++ show (realToFrac n :: Double)
+
+* ``MyMod.10`` => ``"NAT:10"``
+* ``MyMod.(-10)`` => ``"NEGATIVE:-10"``
+* ``MyMod.(10.5)`` => ``"FLOAT:10.5"``
+
+Positive-only literals
+~~~~~~~~~~~~~~~~~~~~~~
+
+The following currently fails to compile with ``-Werror`` enabled:
+
+::
+
+  >>> (-1) :: Natural
+
+  error: [GHC-97441] [-Woverflowed-literals, Werror=overflowed-literals]
+    Literal -1 is negative but Natural only supports positive numbers
+
+However, this check is hardcoded in the compiler for specific types, e.g. ``Natural``, ``Int*``, and ``Word*``. There's no way to enforce this at compile time for custom types, e.g.
+
+::
+
+  -- Invariant: positive
+  newtype UserId = UserId Integer
+
+  UserId (-1) -- works
+
+With ``QualifiedNumbers``, you could define ``fromNumeric`` only for ``Natural`` and not ``Integer``:
+
+::
+
+  -- In UserId module
+  fromNumeric :: Natural -> UserId
+  fromNumeric = UserId
+
+  UserId.123 -- works
+
+  UserId.(-1) -- error: (-1) has type Integer, expected Natural
+
+See https://github.com/ghc-proposals/ghc-proposals/issues/438 for more details.
+
+ByteString
+~~~~~~~~~~
+
+It's a `known issue <https://github.com/haskell/bytestring/issues/140>`_ that ByteString has surprising ``IsString`` behavior, due to ambiguity in how to handle Unicode characters.
+
+With ``QualifiedStrings``, ``bytestring`` could define the following modules:
+
+::
+
+  module Data.ByteString.Qualified.Ascii where
+
+  -- truncates unicode
+  fromString :: String -> ByteString
+  fromString = Char8.pack
+
+  module Data.ByteString.Qualified.Utf8 where
+
+  -- encodes unicode
+  fromString :: String -> ByteString
+  fromString = BS.toStrict . BS.toLazyByteString . BS.stringUtf8
+
+Users would then be forced to decide what behavior they want (and can switch between the two!):
```

Comment:
Why is this better than `bytestring` exposing `fromAscii, fromUtf :: String -> ByteString`?

The issue here seems to be that `ByteString` shouldn't have an `IsString` instance, which seems somewhat orthogonal to this proposal. We don't necessarily need new syntax to allow distinguishing between two conversion functions.

Reply:
Correct; it's not possible to get arbitrary compilation checks here without TH. QualifiedLiterals is purely a desugaring mechanism, it doesn't enable any new capabilities like compile-time checks without TH.

> I'm also unsure how we get from this proposal to working string interpolation.

String interpolation is a completely separate feature. The only reason I bring up the two in relation with each other is that qualified literals introduces syntax that we could reuse to add extra flexibility in string interpolation. But neither proposal can necessarily be implemented in terms of the other